Quip print from NSMenuItem shows up as an empty page

rr-codes opened the pull request:
#### 30ffd1a680bacaf1f0e1b2ecdcc658b3a61d7e90
<pre>
Quip print from NSMenuItem shows up as an empty page
<a href="https://bugs.webkit.org/show_bug.cgi?id=243218">https://bugs.webkit.org/show_bug.cgi?id=243218</a>
rdar://90054391

Reviewed by NOBODY (OOPS!).

When a user pressed Cmd+P, some websites (Quip, Google Docs, etc)
intercept this event and do their own processing before sending a print
event. However, when doing `File &gt; Print`, the default print behavior
is used. This causes discrepancies for how these websites are printed
between the two different methods.

This PR addresses this by synthesizing a Cmd+P event whenever
`File &gt; Print` is invoked or when `Print` is selected from the context
menu, which allows for consistent behavior.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _generateSyntheticCommandP:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::generateSyntheticCommandP):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::generateSyntheticEditingCommand): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::generateSyntheticCommandP):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
</pre>

darinadler commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -974,6 +974,16 @@ bool WebPageProxy::shouldUseBackForwardCache() const
+        return;
+    }
+    
+    sendWithAsyncReply(Messages::WebPage::GenerateSyntheticCommandP(), WTFMove(callbackFunction));
```
This seems to hard-code the command bindings of the web browser.

Is command-P the shortcut for printing worldwide in all the various translated versions of Safari?

I think it’s really sad that the websites hard code a keyboard equivalent, but it seems like this could entirely break printing for some Safari localizations.

We may need to do a more sophisticated fix that always looks like Command-P to the webpage, but then makes sure to actually do a print action once it gets back to the Mac menu system.

Here’s a testing idea: Customize the menu key shortcut for Print, changing it to something like Command-T, using the Keyboard Settings, Keyboard Shortcut, App Shortcuts, and see what happens.

rr-codes replied:
> Is command-P the shortcut for printing worldwide in all the various translated versions of Safari?

Yes, it is (Command+P isn't localized in Safari)/

> Here’s a testing idea: Customize the menu key shortcut for Print, changing it to something like Command-T, using the Keyboard Settings, Keyboard Shortcut, App Shortcuts, and see what happens.

Yup, I already tested this. Specifically, I tested:
* if it works in the DVORAK keyboard layout
* if it works for when the user changes the print command key binding (to `Command+Option+L` for example)
* for a non-Latin system language and keyboard (specifically, Hebrew)

and all of these worked. 

Note that I also have a corresponding Safari PR which is linked in the Radar. The Safari change will trigger the normal print flow if the command+P event was not handled

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -268,6 +268,8 @@ for this property.
 
 - (void)_doAfterNextVisibleContentRectUpdate:(void (^)(void))updateBlock WK_API_AVAILABLE(macos(13.0), ios(16.0));
 
+- (void)_generateSyntheticCommandP:(void (^)(BOOL))completionBlock;
```
This should be marked as `WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))`

rr-codes replied:
do the `TBA`s mean like a future release? If that's the case, will that affect the Safari PR's mergability?

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -47,6 +47,7 @@ messages -> WebPage LegacyReceiver {
     ExecuteEditCommandWithCallback(String name, String argument) -> ()
     KeyEvent(WebKit::WebKeyboardEvent event)
     MouseEvent(WebKit::WebMouseEvent event, std::optional<Vector<WebKit::SandboxExtension::Handle>> sandboxExtensions)
+    GenerateSyntheticCommandP() -> (bool success)
```
Nit - it might also be good to rename `success` here to either `handled` or `prevented`, for clarity?

rr-codes committed:
Quip print from NSMenuItem shows up as an empty page
https://bugs.webkit.org/show_bug.cgi?id=243218
rdar://90054391

Reviewed by NOBODY (OOPS!).

When a user pressed Cmd+P, some websites (Quip, Google Docs, etc)
intercept this event and do their own processing before sending a print
event. However, when doing `File > Print`, the default print behavior
is used. This causes discrepancies for how these websites are printed
between the two different methods.

This PR addresses this by synthesizing a Cmd+P event whenever
`File > Print` is invoked or when `Print` is selected from the context
menu, which allows for consistent behavior.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _generateSyntheticCommandP:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::generateSyntheticCommandP):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::generateSyntheticEditingCommand): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::generateSyntheticCommandP):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
